exercises/markdown/markdown.exs

```diff
@@ -10,9 +10,66 @@ defmodule Markdown do
     iex> Markdown.parse("#Header!\n* __Bold Item__\n* _Italic Item_")
     "<h1>Header!</h1><ul><li><em>Bold Item</em></li><li><i>Italic Item</i></li></ul>"
   """
-
   @spec parse(String.t) :: String.t
-  def parse(markdown) do
+  def parse(m) do
+    patch(Enum.map_join(String.split(m, "\n"), fn(t) -> process(t) end))
+  end
+
+  defp process(t) do
+    if String.starts_with?(t, "#") || String.starts_with?(t, "*") do
+      if String.starts_with?(t, "#") do
+        enclose_with_header_tag(parse_header_md_level(t))
+      else
+        parse_list_md_level(t)
+      end
+    else
+      enclose_with_paragraph_tag(String.split(t))
+    end
+  end
+
+  defp parse_header_md_level(hwt) do
+    [h | t] = String.split(hwt)
+    {to_string(String.length(h)), Enum.join(t, " ")}
+  end
+
+  defp parse_list_md_level(l) do
+    t = String.split(String.trim_leading(l, "* "))
+    "<li>" <> join_words_with_tags(t) <> "</li>"
+  end
+
+  defp enclose_with_header_tag({hl, htl}) do
+    "<h" <> hl <> ">" <> htl <> "</h" <> hl <> ">"
+  end
+
+  defp enclose_with_paragraph_tag(t) do
+    "<p>#{join_words_with_tags(t)}</p>"
+  end
+
+  defp join_words_with_tags(t) do
+    Enum.map_join(t, " ", fn(w) -> replace_md_with_tag(w) end)
+  end
+
+  defp replace_md_with_tag(w) do
+    replace_suffix_md(replace_prefix_md(w))
+  end
+
+  defp replace_prefix_md(w) do
+    cond do
+      w =~ ~r/^[#{"_"}{1}][^#{"_"}+]/ -> String.replace_prefix(w, "_", "<em>")
```

Comment:
Maybe use `replace/4` instead of giving them `replace_prefix/3` and `replace_suffix/3` right away? Also might give you a chance to duplicate the regular expressions?